ftpvl/processors.py

```diff
@@ -0,0 +1,435 @@
+        self._output_col_name = output_col_name
+        
+        assert direction in [1, -1], "direction must be either 1 or -1"
+        self._direction = direction
+
+    def _normalize(self, input_df: pd.DataFrame):
+        """
+        Given a dataframe, find the max value of the input col name and
+        create a new column with the normalized value of each row
+        """
+        if self._direction == 1:
+            max_val = input_df[self._input_col_name].max()
+        else:
+            max_val = input_df[self._input_col_name].min()
+        input_df[self._output_col_name] = input_df[self._input_col_name] / max_val
+        return input_df
+
+    def process(self, input_eval: Evaluation) -> Evaluation:
+        input_df = input_eval.get_df()
+        new_df = input_df.groupby(self._groupby).apply(self._normalize)
+
+        return Evaluation(new_df, input_eval.get_eval_id())
+
+
+class ExpandColumn(Processor):
+    """
+    Processor that turns one column into more than one column by mapping values
+    of a column to multiple values.
+
+    Parameters
+    ----------
+    input_col_name : str
+        the column name to map from
+
+    output_col_names : List[str]
+        the column names to map to
+
+    mapping : dict
+        a dictionary mapping a value to a list of values
+    """
+
+    def __init__(self, input_col_name: str, output_col_names: List[str], mapping: dict):
+
+        self._input_col_name = input_col_name
+        self._output_col_names = output_col_names
+        self._mapping = mapping
+
+    def _expansion(self, input_df: pd.DataFrame):
+        """
+        Given a dataframe that contains only one unique value in
+        _input_col_name, writes new columns as defined by _output_col_names and
+        _mapping and returns the new dataframe
+        """
+        group_value = input_df[self._input_col_name].iloc[0]
+
+        # check for invalid input
+        assert group_value in self._mapping.keys(), f"{group_value} not in mapping"
+        assert len(self._output_col_names) == len(
+            self._mapping[group_value]
+        ), f"{group_value} mapping length does not equal output_col_name length"
+
+        for output_idx in range(len(self._output_col_names)):
+            input_df[self._output_col_names[output_idx]] = self._mapping[group_value][
+                output_idx
+            ]
+
+        return input_df
+
+    def process(self, input_eval: Evaluation) -> Evaluation:
+        input_df = input_eval.get_df()
+        new_df = input_df.groupby(self._input_col_name).apply(self._expansion)
+
+        return Evaluation(new_df, input_eval.get_eval_id())
+
+
+class Reindex(Processor):
+    """
+    Processor that reassigns current columns as indices for easier
+    visualization.
+
+    Reindexing is useful for grouping similar results in the final
+    visualization.
+
+    Parameters
+    ----------
+    reindex_names : List[str]
+        A list of column names to reindex
```

Comment:
I improved the description of Reindex:

```
Reindexing is useful for grouping similar results in the final
    visualization, since each row is identified by a (usually-unique)
    value in the index. Indices are always the first columns, which
    improves readability of the final visualization.
```

The columns "Project", "synthesis_tool", "pr_tool", and "toolchain" are the indices in [this visualization](https://ftpvl.readthedocs.io/en/latest/intro/firststeps.html#visualization).